---
output: github_document
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# Rsearch

## Introduction

Rsearch is a R package designed for handling and analyzing sequence data. Many of its functions are based on the command-line based tool `VSEARCH`. By integrating `VSEARCH` into an R package, `Rsearch` provides significantly utility by making the software more accessible to a broad user base, particularly those who prefer or require an R-based workflow. In addition to the `VSEARCH`-based functions, `Rsearch` includes additional tools to aid in microbiome data analysis, such as data processing functions and visualizations of some of the results from the functions. 

The core idea behind `Rsearch` is to retain the output from `VSEARCH` within R's generic data structures, rather than writing results to files as the original `VSEARCH` functions. By offering the option not to write results to specified files, users can choose to work entirely within R and Rstudio or to export results to files as `VSEARCH` typically does. Keeping all results in R data structures allows users to leverage the power of standard data wrangling and visualization tools familiar to R users. 

Another feature that enhances usability for R users is the consistent return format of the functions. All functions return a single table/data frame unless the user specifies that results should be written to a file. For functions that can return multiple results - such as those handling read pairs with forward and reverse reads - the secondary table is included as an attribute of the primary table. The same approach applies to tables containing statistics from function executions. By ensuring that all functions return only one table, navigating and managing results become more straightforward. 

## Installation

### Prerequisites

For the `Rsearch` package to function on your computer, `VSEARCH` must be installed locally. Please ensure that you are using `VSEARCH` version 2.29.1 or above.

Visit the [`VSEARCH`](https://github.com/torognes/vsearch) page for [instructions](https://github.com/torognes/vsearch?tab=readme-ov-file#download-and-install) on how to download and install the tool. 

#### Installing `VSEARCH` 

The latest release of `VSEARCH`, with corresponding binaries, for installation can be found under [Releases](https://github.com/torognes/vsearch/releases). After downloading the binary, the correct command used to invoke `VSEARCH` on the computer can be set by using the `set_vsearch_executable()` function in `Rsearch` (see [Set correct vsearch executable](#set-correct-vsearch-executable)).

### Installing `Rsearch`

You can install the development version of `Rsearch` from [GitHub](https://github.com/CassandraHjo/Rsearch) by using the `devtools` package from CRAN:

```{r, eval=FALSE}
install.packages("devtools")
devtools::install_github("CassandraHjo/Rsearch")
```

#### Set correct `VSEARCH` executable

In order for most of the functions in `Rsearch` to work, the command to invoke VSEARCH must be set correctly if it is something other than the default `vsearch`.

`set_vsearch_executable` can be used to set the correct command to invoke `VSEARCH` on the computer like this:

```{r eval=FALSE}
# For example, if the vsearch binary (vsearch.exe) is copied to C:/Documents/ on the computer
set_vsearch_executable("C:/Documents/vsearch")
```

## Documentation

### Accessing help within R

Documentation can be accessed directly in the R console. Here are some methods to access help:

* **Function-specific help:**
To get detailed information about a specific function, use the `?` operator followed by the function name. For example, to access help for the `vs_fastq_filter` function:

```{r, eval = FALSE}
?vs_fastq_filter
```

Alternatively, you can use the `help()` function:

```{r, eval = FALSE}
help(vs_fstq_filter)
```

* **Package-wide help**

To get an overview of the `Rsearch` package an its available functions, use:

```{r, eval = FALSE}
# library(Rsearch)
help(package = "Rsearch")
```

## Usage

Additional usage examples can be found in the [documentation](#documentation) for each individual function.

### Example: Filter paired-end reads based on quality

```{r eval=FALSE}
library(Rsearch)

# Define FASTQ files
fastq_input <- "R1_sample1.fq"
reverse <- "R2_sample1.fq"

# Define other arguments
output_format <- "fastq"
fastq_maxee_rate <- 0.01
minlen <- 0

# Execute filtering, with tibble as output
filt_seqs <- vs_fastq_filter(fastq_input = fastq_input,
                             reverse = reverse,
                             output_format = output_format,
                             fastq_maxee_rate = fastq_maxee_rate,
                             minlen = minlen)

# Extract tibbles with filtered sequences
R1_filt <- filt_seqs
R2_filt <- attr(filt_seqs, "filt_reverse")

# Extract filtering statistics
statistics <- attr(filt_seqs, "statistics")
```

## Contributors

The main contributors to `Rsearch`:

* Cassandra Hjortdahl <cassandra.hjortdahl@nmbu.no> (Coding, testing, documentation, maintaining)
* Lars Snipen <lars.snipen@nmbu.no> (Coding, testing)
* Torbjørn Rognes <torognes@ifi.uio.no> (Coding, testing)
* Hilde Vinje <hilde.vinje@nmbu.no> (Coding, testing)

## Citing Rsearch

Please cite the following publication if you use `Rsearch`:

xxx

Please note that citing any of the underlying algorithms, e.g. `VSEARCH`, may also be appropriate. 

## References

* Rognes T, Flouri T, Nichols B, Quince C, Mahé F (2016)
**VSEARCH: a versatile open source tool for metagenomics.**
*PeerJ* 4:e2584.
doi: [10.7717/peerj.2584](https://doi.org/10.7717/peerj.2584)
